ui/src/vertical-geometry/snapped-point.ts

```diff
@@ -0,0 +1,200 @@
+    getTrackAddressAtSingleIndex,
+    getTrackMeterPairAroundIndex,
+    TrackMeterIndex,
+} from 'vertical-geometry/track-meter-index';
+import { StationPoint, VerticalGeometryItem } from 'geometry/geometry-model';
+import { filterNotEmpty, minimumIndexBy } from 'utils/array-utils';
+import { Coordinates, heightToY, mToX, xToM } from 'vertical-geometry/coordinates';
+import { TrackKmHeights } from 'geometry/geometry-api';
+import { approximateHeightAt } from 'vertical-geometry/util';
+
+export type SnapTarget =
+    | 'trackMeterRulerTick'
+    | 'segmentBoundary'
+    | 'geometryElement'
+    | 'didNotSnap';
+
+export interface SnappedPoint {
+    snapTarget: SnapTarget;
+    m: number;
+    xPositionPx: number;
+    yPositionPx: number;
+    height: number;
+    address: TrackMeter;
+}
+
+function getSnapOverRuler(
+    xCoordinateM: number,
+    trackKmHeights: TrackKmHeights[],
+    onScreen: (snappedM: number) => boolean,
+    approximatedPoint: (
+        maybeApproximateM: number,
+    ) => null | { address: TrackMeter | null; height: number | null },
+): { address: TrackMeter | null; m: number; snapTarget: SnapTarget; height: number | null } {
+    const closest = closestRulerTickM(xCoordinateM, trackKmHeights);
+    if (closest == null || !onScreen(closest.m)) {
+        const approximated = approximatedPoint(xCoordinateM);
+        return {
+            snapTarget: 'didNotSnap',
+            m: xCoordinateM,
+            height: approximated?.height ?? null,
+            address: approximated?.address ?? null,
+        };
+    } else {
+        const address = getTrackAddressAtSingleIndex(closest.index, trackKmHeights);
+        return {
+            snapTarget:
+                address.meters === Math.floor(address.meters)
+                    ? 'trackMeterRulerTick'
+                    : 'segmentBoundary',
+            m: closest.m,
+            address: getTrackAddressAtSingleIndex(closest.index, trackKmHeights),
+            height: approximateHeightAt(closest.m, closest.interval, trackKmHeights),
+        };
+    }
+}
+
+function getSnapOverChart(
+    xCoordinateM: number,
+    geometry: VerticalGeometryItem[],
+    withinSnapDistance: (snappedM: number) => boolean,
+    approximatedPoint: (
+        maybeApproximateM: number,
+    ) => null | { address: TrackMeter | null; height: number | null },
+    drawTangents: boolean,
+): { address: TrackMeter | null; m: number; snapTarget: SnapTarget; height: number | null } {
+    const closest = closestGeometrySnapPoint(xCoordinateM, geometry, drawTangents);
+    if (closest == null || !withinSnapDistance(closest.m)) {
+        const approximated = approximatedPoint(xCoordinateM);
+        return {
+            snapTarget: 'didNotSnap',
+            m: xCoordinateM,
+            height: approximated?.height ?? null,
+            address: approximated?.address ?? null,
+        };
+    }
+
+    const height =
+        closest.type === 'intersectionPoint'
```

Comment:
Tämä logiikka että juuri PVI-pisteiden kohdalla näytetäänkin raiteen korkeutta, mutta muiden geometriaelementtien kohdalla elementin omaa korkeutta, on minun vähän tuulesta tempaama, sillä perusteella että pyöristyksen alku/loppuhan on juurikin raiteella (paitsi jos se ei olekaan se geometriaelementti, joka tässä määrää raiteen korkeuden), kun taas PVI-pisteellä se on tangenttipiste; ja kun tangenttipisteen korkeus näkyy siinä heti vieressä, niin tällä tavallahan saadaan enemmän tietoa näkyviin. Pitää vielä Karilta kysyä, onko tämä nyt kuinka outo ajatus tehdä näin.